File: backend/models/answer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
答案之书应用 - 答案模型

这个模块定义了答案数据模型。
"""

from datetime import datetime
from . import db


class Answer(db.Model):
    """答案模型"""
    __tablename__ = 'answers'
    
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    content = db.Column(db.String(500), nullable=False, comment='答案内容')
    category_id = db.Column(db.Integer, db.ForeignKey('categories.id'), nullable=True, comment='分类ID')
    created_at = db.Column(db.DateTime, default=datetime.now, comment='创建时间')
    updated_at = db.Column(db.DateTime, default=datetime.now, onupdate=datetime.now, comment='更新时间')
    is_active = db.Column(db.Boolean, default=True, comment='是否激活')
    
    # 关联分类
    category = db.relationship('Category', backref=db.backref('answers', lazy='dynamic'))
    
    def __repr__(self):
        return f'<Answer: {self.content[:20]}...>'
    
    def to_dict(self):
        """将模型转换为字典"""
        return {
            'id': self.id,
            'content': self.content,
            'category_id': self.category_id,
            'category_name': self.category.name if self.category else None,
            'created_at': self.created_at.strftime('%Y-%m-%d %H:%M:%S'),
            'updated_at': self.updated_at.strftime('%Y-%m-%d %H:%M:%S'),
            'is_active': self.is_active
        }
    
    @classmethod
    def get_random(cls):
        """获取随机答案"""
        from sqlalchemy.sql.expression import func
        
        # 使用数据库的随机排序功能获取一个随机答案
        answer = cls.query.filter_by(is_active=True).order_by(func.random()).first()
        return answer

File: frontend/lib/models/answer.dart
class Answer {
  final int id;
  final String content;
  final int? categoryId;
  final String? categoryName;
  final String? createdAt;
  final String? updatedAt;
  final bool isActive;
  
  Answer({
    required this.id,
    required this.content,
    this.categoryId,
    this.categoryName,
    this.createdAt,
    this.updatedAt,
    this.isActive = true,
  });
  
  // 从JSON创建答案对象
  factory Answer.fromJson(Map<String, dynamic> json) {
    return Answer(
      id: json['id'],
      content: json['content'],
      categoryId: json['category_id'],
      categoryName: json['category_name'],
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
      isActive: json['is_active'] ?? true,
    );
  }
  
  // 将答案对象转换为JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'content': content,
      'category_id': categoryId,
      'category_name': categoryName,
      'created_at': createdAt,
      'updated_at': updatedAt,
      'is_active': isActive,
    };
  }
  
  // 创建答案对象的副本
  Answer copyWith({
    int? id,
    String? content,
    int? categoryId,
    String? categoryName,
    String? createdAt,
    String? updatedAt,
    bool? isActive,
  }) {
    return Answer(
      id: id ?? this.id,
      content: content ?? this.content,
      categoryId: categoryId ?? this.categoryId,
      categoryName: categoryName ?? this.categoryName,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isActive: isActive ?? this.isActive,
    );
  }
}